src/network/peers.rs

```diff
@@ -595,29 +602,23 @@ impl<N: Network, E: Environment> Peers<N, E> {
 
                         true
                     }
-                    Message::UnconfirmedBlock(_, _, ref mut data) => {
-                        let block = if let Data::Object(block) = data {
-                            block
-                        } else {
-                            panic!("Logic error: the block shouldn't have been serialized yet.");
-                        };
-
+                    Message::UnconfirmedBlock(block_height, block_hash, ref mut data) => {
                         // Lock seen_outbound_blocks for further processing.
                         let mut seen_outbound_blocks = self.seen_outbound_blocks.write().await;
 
                         // Retrieve the last seen timestamp of this block for this peer.
                         let seen_blocks = seen_outbound_blocks.entry(peer).or_insert_with(Default::default);
-                        let last_seen = seen_blocks.entry(block.hash()).or_insert(SystemTime::UNIX_EPOCH);
+                        let last_seen = seen_blocks.entry(block_hash).or_insert(SystemTime::UNIX_EPOCH);
                         let is_ready_to_send = last_seen.elapsed().unwrap().as_secs() > E::RADIO_SILENCE_IN_SECS;
 
                         // Update the timestamp for the peer and sent block.
-                        seen_blocks.insert(block.hash(), SystemTime::now());
+                        seen_blocks.insert(block_hash, SystemTime::now());
                         // Report the unconfirmed block height.
                         if is_ready_to_send {
-                            trace!("Preparing to send 'UnconfirmedBlock {}' to {}", block.height(), peer);
+                            trace!("Preparing to send 'UnconfirmedBlock {}' to {}", block_height, peer);
                         }
 
-                        // Perform non-blocking serialization of the block.
+                        // Perform non-blocking serialization of the block (if it hasn't been serialized yet).
```

Comment:
Please revert this until a more thorough assessment is made.

The implication is that a malicious block from modified code can admit this change and poison internal `last_seen` filters if it is incorrect.

A simple fix is to check that `block_height == block.height()` and `block_hash == block.hash()` here, however this results in redundant checks. I can take a look at this change more closely later on to ensure we have it enforced at the lowest level possible.